pkg/retriever/graphsyncretriever.go

```diff
@@ -148,12 +166,32 @@ func (r *graphsyncRetrieval) RetrieveFromAsyncCandidates(asyncCandidates types.I
 				return
 			}
 			for _, candidate := range candidates {
-				candidate := candidate
-				waitGroup.Add(1)
-				go func() {
-					defer waitGroup.Done()
-					runRetrievalCandidate(ctx, r.GraphSyncRetriever, r.request, r.Client, r.request.LinkSystem, retrieval, phaseStartTime, candidate)
-				}()
+				// Check if we already started a retrieval for this candidate
+				r.candidateMetdataMut.RLock()
+				currMetadata, seenCandidate := r.candidateMetadata[candidate.MinerPeer.ID]
+				r.candidateMetdataMut.RUnlock()
+				if seenCandidate {
+					// Don't start a new retrieval, but update the metadata if it's more favorable
+					newMetadata := candidate.Metadata.Get(multicodec.TransportGraphsyncFilecoinv1).(*metadata.GraphsyncFilecoinV1)
+					if metadataCompare(newMetadata, currMetadata) {
```

Comment:
I really like reusing the metadataCompare func here -- that's a great design!

Reply:
Thank you!